core/amber/src/test/scala/edu/uci/ics/amber/engine/e2e/BatchSizePropagationSpec.scala

```diff
@@ -0,0 +1,162 @@
+  }
+
+  "Engine" should "propagate the correct batch size into each WorkflowContext" in {
+    val expectedBatchSize = 1
+
+    val customWorkflowSettings = WorkflowSettings(dataTransferBatchSize = expectedBatchSize)
+
+    val context = new WorkflowContext(workflowSettings = customWorkflowSettings)
+
+    val headerlessCsvOpDesc = TestOperators.headerlessSmallCsvScanOpDesc()
+    val sink = TestOperators.sinkOpDesc()
+
+    val workflowCompiler = new WorkflowCompiler(
+      context
+    )
+
+    val workflow = workflowCompiler.compile(
+      LogicalPlanPojo(
+        List(headerlessCsvOpDesc, sink),
+        List(
+          LogicalLink(
+            headerlessCsvOpDesc.operatorIdentifier,
+            PortIdentity(),
+            sink.operatorIdentifier,
+            PortIdentity()
+          )
+        ),
+        List(),
+        List()
+      ),
+      resultStorage,
+      new ExecutionStateStore()
+    )
+
+    executeWorkflow(workflow)
+
+    val workflowScheduler = new WorkflowScheduler(context, resultStorage)
+    workflowScheduler.updateSchedule(workflow.physicalPlan)
+
+    var nextRegions = workflowScheduler.getNextRegions
+    while (nextRegions.nonEmpty) {
+      nextRegions.foreach { region =>
+        region.resourceConfig.foreach { resourceConfig =>
+          resourceConfig.linkConfigs.foreach {
+            case (_, linkConfig) =>
+              val partitioning = linkConfig.partitioning
+              partitioning match {
+                case oneToOne: OneToOnePartitioning =>
+                  assert(
+                    oneToOne.batchSize == expectedBatchSize,
+                    s"Batch size mismatch: ${oneToOne.batchSize} != $expectedBatchSize"
+                  )
+
+                case roundRobin: RoundRobinPartitioning =>
+                  assert(
+                    roundRobin.batchSize == expectedBatchSize,
+                    s"Batch size mismatch: ${roundRobin.batchSize} != $expectedBatchSize"
+                  )
+
+                case hashBased: HashBasedShufflePartitioning =>
+                  assert(
+                    hashBased.batchSize == expectedBatchSize,
+                    s"Batch size mismatch: ${hashBased.batchSize} != $expectedBatchSize"
+                  )
+
+                case rangeBased: RangeBasedShufflePartitioning =>
+                  assert(
+                    rangeBased.batchSize == expectedBatchSize,
+                    s"Batch size mismatch: ${rangeBased.batchSize} != $expectedBatchSize"
+                  )
+
+                case broadcast: BroadcastPartitioning =>
+                  assert(
+                    broadcast.batchSize == expectedBatchSize,
+                    s"Batch size mismatch: ${broadcast.batchSize} != $expectedBatchSize"
+                  )
+
+                case _ =>
+                  println("Unknown partitioning type")
+              }
+          }
+        }
+      }
+      nextRegions = workflowScheduler.getNextRegions
```

Comment:
try more workflows. this workflow is a bit simple. it only contains a source and a sink.